File: valid_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        lb, ub = float("-inf"), float("inf")
        return self.dfs(root, lb, ub)

    def dfs(self, root, lb, ub):
        if not root:
            return True

        if (root.val < ub and root.val > lb):
            return self.dfs(root.left, lb, root.val) and self.dfs(root.right, root.val, ub)
        else:
            return False

File: validBinarySearchTree.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
  /**
   * @param {TreeNode} root
   * @return {boolean}
   */
  isValidBST(root) {
    let lb = -Infinity;
    let ub = Infinity;
    const dfs = (root, lb, ub) => {
      if (!root) {
        return true;
      }

      if (root.val > lb && root.val < ub) {
        return dfs(root.left, lb, root.val) && dfs(root.right, root.val, ub);
      } else {
        return false;
      }
    };

    return dfs(root, lb, ub);
  }
}
